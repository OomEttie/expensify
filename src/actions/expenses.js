const uuidV4 = require('uuid/v4');
import database from '../firebase/firebase';

//get the correct expenses db ref path
let expensesRef = uid => {
  return 'users/' + uid + '/expenses';
};

//
// ADD_EXPENSES
//
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    
    return database
      .ref(dbPath)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//
// REMOVE_EXPENSE
//
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id: id
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(`${dbPath}/${id}`)
      .remove()
      .then(data => {
        dispatch(removeExpense({ id }));
      });
  };
};

//
// EDIT_EXPENSE
//
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(`${dbPath}/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

//
// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(dbPath)
      .once('value')
      .then(snapshot => {
        const expenses = [];

        snapshot.forEach(snapshotChild => {
          expenses.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch(e => {});
  };
};
