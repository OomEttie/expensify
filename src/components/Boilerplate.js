import React from 'react';
import Header from './Header';

export default class BoilerplateApp extends React.Component {
  state = {
  };

  //FUNCTIONS
  render() {
    return (
      <div>
        <Header />
      </div>
    );
  }
}
