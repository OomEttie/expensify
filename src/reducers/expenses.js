const expensesReducerDefaultState = [];

export const expensesReducer = (
  state = expensesReducerDefaultState,
  action
) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    // return state.concat(action.expense);
    case 'REMOVE_EXPENSE':
      return state.filter(data => data.id != action.id);
    case 'EDIT_EXPENSE': {
      // const expenseIndex = state.findIndex(data => data.id == action.id);
      // return { ...state[expenseIndex], amount: action.amount };
      return state.map(expense => {
        // console.log('expense-', expense);
        // console.log('EDIT_EXPENSE action-', action.id);
        if (expense.id == action.id) {
          return { ...expense, ...action.updates };
        } else {
          return expense;
        }
      });
    }
    case 'SET_EXPENSES':
      return action.expenses;
    default:
      return state;
  }
};
