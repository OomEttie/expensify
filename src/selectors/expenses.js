import moment from 'moment';

export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      // typeof startDate != 'number' || expense.createdAt >= startDate;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      // typeof endDate != 'number' || expense.createdAt <= endDate;
      const textMatch =
        expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy == 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy == 'amount') {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};
