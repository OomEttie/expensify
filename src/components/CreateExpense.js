import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// onSubmit = expense => {
//   console.log('onSubmit-', expense);
// };

export class CreateExpense extends React.Component {
  onSubmit = expense => {
    // props.dispatch(addExpense(expense));
    this.props.createExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    createExpense: expense => {
      dispatch(startAddExpense(expense));
    }
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(CreateExpense);
