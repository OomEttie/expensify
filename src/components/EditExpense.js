import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  onSubmit = expense => {
    // props.dispatch(addExpense(expense));
    this.props.startEditExpense(this.props.expense.id, { ...expense });
    this.props.history.push('/');
  };

  onClick = () => {
    // console.log(dispatch);
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button__warning" onClick={this.onClick}>REMOVE</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  // console.log('mapStateToProps-props->', props)
  return {
    expense: state.expenses.find(expense => {
      return expense.id == props.match.params.id;
    })
  };
};

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, updates) => {
      dispatch(startEditExpense(id, updates));
    },
    startRemoveExpense: id => {
      dispatch(startRemoveExpense(id));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
