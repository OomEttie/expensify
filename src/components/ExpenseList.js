import React from 'react';
import { connect } from 'react-redux';
import { ExpenseListItem } from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';

// <h1>Expense List</h1>
export const ExpenseList = props => (
  <div>
    {props.expenses.length == 0 ? (
      <p>No Expenses</p>
    ) : (
      props.expenses.map(expense => {
        // console.log(expense);
        // const Temp = () => (
        //   <div>
        //     <p>Amount: </p>
        //     <p>Created At: </p>
        //   </div>
        // )
        // return <Temp />;
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

// export default ExpenseList;
export default ConnectedExpenseList;
