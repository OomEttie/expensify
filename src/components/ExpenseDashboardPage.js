import React from 'react';
import { connect } from 'react-redux';

import ExpenseSummaryPage from './ExpenseSummaryPage';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import { getVisibleExpenses } from '../selectors/expenses';
import { getTotalExpensesAmount } from '../selectors/expenses-total';

export class ExpenseDashboardPage extends React.Component {
  render() {
    return (
      <div>
        <ExpenseSummaryPage />
        <ExpenseListFilters />
        <ExpenseList />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseDashboardPage);

// export default ExpenseDashboardPage;
// export default ConnectedExpenseDashboardPage;
