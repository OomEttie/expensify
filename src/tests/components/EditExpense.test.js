import React from 'react';
import { shallow } from 'enzyme';

import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let wrapper, history, removeExpense, editExpense;

beforeEach(() => {
  removeExpense = jest.fn();
  editExpense = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(
    <EditExpense
      expense={expenses[1]}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  );
});

test('should render EditExpense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('shoud handle EditExpense onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('shoud handle EditExpense onClick(onremove)', () => {
  wrapper.find('button').prop('onClick')();

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});
