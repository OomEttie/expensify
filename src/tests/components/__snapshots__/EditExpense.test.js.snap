// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 109500,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "1",
          "note": "2 Rent note",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button__warning"
      onClick={[Function]}
    >
      REMOVE
    </button>
  </div>
</div>
`;
